### PR TITLE
[Merged by Bors] -  feat(Mathlib.NumberTheory.FLT.Three): add various results

### PR DESCRIPTION
**Pull request by riccardobrasca**
We add various results needed to prove flt3.

From the flt3 project at LFTCM2024.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/Cyclotomic/Rat.lean [hunk 1]
```diff
@@ -563,10 +576,12 @@ variable (K p k)
 /-- We compute the absolute discriminant of a `p ^ k`-th cyclotomic field.
   Beware that in the cases `p ^ k = 1` and `p ^ k = 2` the formula uses `1 / 2 = 0` and `0 - 1 = 0`.
   See also the results below. -/
-theorem absdiscr_prime_pow [NumberField K] [IsCyclotomicExtension {p ^ k} ℚ K] :
+theorem absdiscr_prime_pow[IsCyclotomicExtension {p ^ k} ℚ K] :
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Cyclotomic/Rat.lean:579]:
```suggestion
theorem absdiscr_prime_pow [IsCyclotomicExtension {p ^ k} ℚ K] :
```

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 2]
```diff
@@ -198,6 +197,154 @@ lemma Solution.exists_minimal : ∃ (S₁ : Solution hζ), S₁.isMinimal := by
...
+both congruent to `-1`.  -/
+lemma a_cube_b_cube_congr_one_or_neg_one :
+    λ ^ 4 ∣ S'.a ^ 3 - 1 ∧ λ ^ 4 ∣ S'.b ^ 3 + 1 ∨ λ ^ 4 ∣ S'.a ^ 3 + 1 ∧ λ ^ 4 ∣ S'.b ^ 3 - 1 := by
+  obtain ⟨z, hz⟩ := S'.hcdvd
+  rcases lambda_pow_four_dvd_cube_sub_one_or_add_one_of_lambda_not_dvd hζ S'.ha with
+    (⟨x, hx⟩ | ⟨x, hx⟩) <;>
+  rcases lambda_pow_four_dvd_cube_sub_one_or_add_one_of_lambda_not_dvd hζ S'.hb with
+    (⟨y, hy⟩ | ⟨y, hy⟩)
+  · exfalso
+    replace hζ : IsPrimitiveRoot ζ ((3 : ℕ+) ^ 1) := hζ
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:210]: Maybe better as
```suggestion
    replace hζ : IsPrimitiveRoot ζ ((3 : ℕ+) ^ 1) := by rwa [pow_one]
```

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 3]
```diff
@@ -198,6 +197,154 @@ lemma Solution.exists_minimal : ∃ (S₁ : Solution hζ), S₁.isMinimal := by
...
+    symm
+    calc _ = S'.u * (λ * z) ^ 3 - λ ^ 4 * x - λ ^ 4 * y := by ring
+    _ = (S'.a ^ 3 + S'.b ^ 3) - (S'.a ^ 3 - 1) - (S'.b ^ 3 - 1) := by rw [← hx, ← hy, ← hz, ← S'.H]
+    _ = 2 := by ring
+  · left
+    exact ⟨⟨x, hx⟩, ⟨y, hy⟩⟩
+  · right
+    exact ⟨⟨x, hx⟩, ⟨y, hy⟩⟩
+  · exfalso
+    replace hζ : IsPrimitiveRoot ζ ((3 : ℕ+) ^ 1) := hζ
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:221]: Ditto
```suggestion
    replace hζ : IsPrimitiveRoot ζ ((3 : ℕ+) ^ 1) := by rwa [pow_one]
```

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 4]
```diff
@@ -198,6 +197,154 @@ lemma Solution.exists_minimal : ∃ (S₁ : Solution hζ), S₁.isMinimal := by
...
+/-- Given `S' : Solution'`, we have that `λ ^ 4` divides `S'.c ^ 3`. -/
+lemma lambda_pow_four_dvd_c_cube : λ ^ 4 ∣ S'.c ^ 3 := by
+  rcases a_cube_b_cube_congr_one_or_neg_one S' with
+    (⟨⟨x, hx⟩, ⟨y, hy⟩⟩ | ⟨⟨x, hx⟩, ⟨y, hy⟩⟩) <;> {
+  refine ⟨S'.u⁻¹ * (x + y), ?_⟩
+  symm
+  calc _ = S'.u⁻¹ * (λ ^ 4 * x + λ ^ 4 * y) := by ring
+  _ = S'.u⁻¹ * (S'.a ^ 3 + S'.b ^ 3) := by rw [← hx, ← hy]; ring
+  _ = S'.u⁻¹ * (S'.u * S'.c ^ 3) := by rw [S'.H]
+  _ = S'.c ^ 3 := by simp }
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:237]: I don't think we use brackets like that anymore
```suggestion
    (⟨⟨x, hx⟩, ⟨y, hy⟩⟩ | ⟨⟨x, hx⟩, ⟨y, hy⟩⟩) <;>
  · refine ⟨S'.u⁻¹ * (x + y), ?_⟩
    symm
    calc _ = S'.u⁻¹ * (λ ^ 4 * x + λ ^ 4 * y) := by ring
    _ = S'.u⁻¹ * (S'.a ^ 3 + S'.b ^ 3) := by rw [← hx, ← hy]; ring
    _ = S'.u⁻¹ * (S'.u * S'.c ^ 3) := by rw [S'.H]
    _ = S'.c ^ 3 := by simp
```

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 5]
```diff
@@ -198,6 +197,154 @@ lemma Solution.exists_minimal : ∃ (S₁ : Solution hζ), S₁.isMinimal := by
...
+
+/-- Given `S' : Solution'`, we have that `λ ^ 2` divides `S'.c`. -/
+lemma lambda_sq_dvd_c : λ ^ 2 ∣ S'.c := by
+  classical
+  have  hm := S'.multiplicity_lambda_c_finite
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:242]:
```suggestion
  have hm := S'.multiplicity_lambda_c_finite
```

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 6]
```diff
@@ -198,6 +197,154 @@ lemma Solution.exists_minimal : ∃ (S₁ : Solution hζ), S₁.isMinimal := by
...
+  suffices 2 ≤ (multiplicity ((hζ.toInteger - 1)) S'.c).get hm by
+    · obtain ⟨x, hx⟩ := multiplicity.pow_multiplicity_dvd hm
+      refine ⟨λ ^ ((multiplicity ((hζ.toInteger - 1)) S'.c).get hm - 2) * x, ?_⟩
+      rw [← mul_assoc, ← pow_add]
+      convert hx using 3
+      simp [this]
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:248]:
```suggestion
    obtain ⟨x, hx⟩ := multiplicity.pow_multiplicity_dvd hm
    refine ⟨λ ^ ((multiplicity ((hζ.toInteger - 1)) S'.c).get hm - 2) * x, ?_⟩
    rw [← mul_assoc, ← pow_add]
    convert hx using 3
    simp [this]
```

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 7]
```diff
@@ -198,6 +197,154 @@ lemma Solution.exists_minimal : ∃ (S₁ : Solution hζ), S₁.isMinimal := by
...
+  have := lambda_pow_four_dvd_c_cube S'
+  have hm1 :(multiplicity (hζ.toInteger - 1) (S'.c ^ 3)).get
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:250]:
```suggestion
  have hm1 : (multiplicity (hζ.toInteger - 1) (S'.c ^ 3)).get
```

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 8]
```diff
@@ -198,6 +197,154 @@ lemma Solution.exists_minimal : ∃ (S₁ : Solution hζ), S₁.isMinimal := by
...
+lemma a_cube_add_b_cube_eq_mul :
+    S'.a ^ 3 + S'.b ^ 3 = (S'.a + S'.b) * (S'.a + η * S'.b) * (S'.a + η ^ 2 * S'.b) := by
+  have := hζ.isRoot_cyclotomic (by decide)
+  simp only [PNat.val_ofNat, Polynomial.cyclotomic_three, Polynomial.IsRoot.def,
+    Polynomial.eval_add, Polynomial.eval_pow, Polynomial.eval_X, Polynomial.eval_one] at this
+  symm
+  calc _ = S'.a^3+S'.a^2*S'.b*(η^2+η+1)+S'.a*S'.b^2*(η^2+η+η^3)+η^3*S'.b^3 := by ring
+  _ = S'.a^3+S'.a^2*S'.b*(η^2+η+1)+S'.a*S'.b^2*(η^2+η+1)+S'.b^3 := by
+    norm_cast
+    simp [show hζ.toInteger ^ 3 = 1 from hζ.toInteger_isPrimitiveRoot.pow_eq_one]
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:276]: Did you not add a lemma for this?

**Review by Ruben-VandeVelde — APPROVED** [review]
lgtm

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**